src/transport/PeerMessageCounter.h

```diff
@@ -144,14 +161,27 @@ class PeerMessageCounter
      */
     void Commit(uint32_t counter)
     {
+        uint32_t offset = mSynced.mMaxCounter - counter;
         if (counter <= mSynced.mMaxCounter)
         {
-            uint32_t offset = mSynced.mMaxCounter - counter;
+            if (mIsGroup && (offset > (UINT32_MAX / 2)))
+            {
+                // Different behaviour since Group Counter can roll over
+                // spec 4.5.1.3
+
+                // Possible roll over, set the offset to the smallest of the two
+                uint32_t secondOffset = counter - mSynced.mMaxCounter;
```

Comment:
Again, I don't think this computation works correctly in the simple "we just overflowed" case.

I _think_, again, that we know at all the Commit() callsites what sort of counter we have, and we can do `CommitRolloverAllowed` and `CommitRolloverDisallowed` or something....  Trying to shoehorn both behaviors into the same function is likely to be very hard to understand and error-prone.

Reply:
reworked.